ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/Jumping.hs

```diff
@@ -0,0 +1,294 @@
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (Instruction blk)
+
+data JumpResult blk
+  = AcceptedJump !(Point blk)
+  | RejectedJump !(Point blk)
+  deriving (Generic)
+
+deriving instance (HasHeader blk, Eq (Header blk)) => Eq (JumpResult blk)
+deriving instance (HasHeader blk, Show (Header blk)) => Show (JumpResult blk)
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (JumpResult blk)
+
+nextInstruction ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk, HasHeader (Header blk)
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  STM m (Instruction blk)
+nextInstruction handlesVar peer =
+  (cschJumping . (Map.! peer) <$> readTVar handlesVar) >>= \case
+    Dynamo lastJumpSlot -> maybeSetNextJump lastJumpSlot >> pure RunNormally
+    Objector _ -> pure RunNormally
+    Jumper nextJumpVar _ _ ->
+      readTVar nextJumpVar >>= \case
+        Nothing -> retry
+        Just fragment -> do
+          writeTVar nextJumpVar Nothing
+          pure $ JumpTo fragment
+  where
+    -- | When the tip of the candidate fragment is 'jumpSize' slots younger than
+    -- the last jump, set jumpers to jump to it.
+    maybeSetNextJump lastJumpSlot = do
+      dynamoFragment <- fromJust <$> getDynamoFragment handlesVar
+      when (succWithOrigin (headSlot dynamoFragment) >= succWithOrigin lastJumpSlot + jumpSize) $ do
+        handles <- readTVar handlesVar
+        forM_ (map cschJumping $ Map.elems handles) $ \case
+          Jumper nextJumpVar _ Happy -> writeTVar nextJumpVar $ Just $! castPoint (headPoint dynamoFragment)
+          _ -> pure ()
+
+setJumpingState ::
+  ( MonadSTM m,
+    Ord peer
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  ChainSyncJumpingState m blk ->
+  STM m ()
+setJumpingState handlesVar peer cschJumping =
+  modifyTVar handlesVar $ Map.update (\handle -> Just $! handle {cschJumping}) peer
+
+processJumpResult ::
+  ( MonadSTM m,
+    Ord peer, HasHeader blk, HasHeader (Header blk)
+  ) =>
+  StrictTVar m (Map peer (ChainSyncClientHandle m blk)) ->
+  peer ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult handlesVar peer jumpResult =
+  (cschJumping . (Map.! peer) <$> readTVar handlesVar) >>= \case
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment handlesVar
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
```

Comment:
This `fromJust` call might fail if the dynamo changes right before performing the jump, or if the dynamo does a rollback.

A possible fix is to store the dynamoFragment together with the jump offer, so the candidate fragment can be updated afterwards regardless of what happens with the dynamo.